src/web/server/app/services/api.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  SELECT_CAMERAID,
+  SELECT_USERID_CAMERAID,
+  SELECT_CAMERAID_USERID,
+  SELECT_LIVE_CAMERAID,
+  SELECT_STREAMID_METADATA,
+  SELECT_CAMERAID_STREAMID,
+  SELECT_CAMERAID_STREAMID_SINGLE,
+  INSERT_CAMERAID_STREAMID,
+  INSERT_USERID_CAMERAID,
+  INSERT_CAMERAID_USERID,
+  INSERT_CAMERAID_LIVE,
+} from '../utils/queries';
+
+import { client } from '../utils/client';
+
+/**
+ * Retrieve streamIds from database
+ * @param cameraId - The cameraId that spawned the streamIds
+ */
+export const retrieveStreamIds = async (cameraId: string) => {
+  return client.execute(SELECT_CAMERAID_STREAMID, [cameraId], {
+    prepare: true,
+  });
+};
+
+/**
+ * Retrieve streamId from camera if live from database
+ * @param cameraId - The cameraId that may be live
+ */
+export const retrieveLiveStreamId = async (cameraId: string) => {
+  const result = await client.execute(SELECT_LIVE_CAMERAID, [cameraId], {
+    prepare: true,
+  });
+  if (result.rows.length === 0 || result.rows[0]['live'] !== true) {
+    return undefined;
+  }
+
+  return client.execute(SELECT_CAMERAID_STREAMID_SINGLE, [cameraId], {
+    prepare: true,
+  });
+};
+
+/**
+ * Retrieve all streamId from camera if live belonging to user
+ * @param userId - userId of user
+ */
+export const retrieveLiveCameraStreamIds = async (userId: string) => {
+  const camerasResult = await retrieveCameraIds(userId);
+  if (camerasResult === undefined || camerasResult.rows.length === 0) {
+    return undefined;
+  }
+  const liveCameras = camerasResult.rows.reduce(
+    async (oldCollection: Promise<Record<string, string>>, row: any) => {
+      const collection = await oldCollection;
+      const liveStreamIdResult = await retrieveLiveStreamId(row['camera_id']);
+      if (
+        liveStreamIdResult !== undefined &&
+        liveStreamIdResult.rows.length > 0
+      ) {
+        collection[row['camera_id']] = liveStreamIdResult.rows[0]['stream_id'];
+      }
+      return collection;
+    },
+    new Promise((resolve, reject) => resolve({}))
+  );
+  return liveCameras;
+};
+
+/**
+ * Store streamId in database
+ * @param cameraId - The cameraId of the camera that streamed
+ * @param streamId - The streamId
+ */
+export const storeStreamId = async (cameraId: string, streamId: string) => {
+  const params = [cameraId, streamId, Date.now()];
+  const result = await client.execute(SELECT_CAMERAID_STREAMID, [cameraId], {
+    prepare: true,
+  });
+  if (!result.rows.some(element => element['stream_id'] === streamId)) {
+    await client.execute(INSERT_CAMERAID_STREAMID, params, { prepare: true });
+  }
+};
+
+/**
+ * Update if camera is live in database
+ * @param cameraId - The cameraId of the camera that streamed
+ * @param live - Whether the camera started or stopped streaming
+ */
+export const updateCameraLive = async (cameraId: string, live: boolean) => {
+  const params = [cameraId, live];
+  await client.execute(INSERT_CAMERAID_LIVE, params, { prepare: true });
```

Comment:
async and await does nothing in this method

Reply:
It seems like I need it if a subsequent code relies on the execution of this function. 
I.e. in my test that calls this function, I mark a cameraId's live attribute to be true and retrieve its live streamId, then I mark its live attribute to be false, and retrieve its streamId again (which should be undefined since the camera is no longer live). With async/await it correctly returns undefined, but without async/await, it sometimes returns a streamId.

Reply:
I suggest that you should return some value indicating that it was successfully executed. I understand how await blocks the call, but if client.execute runs into a problem, then rip

Reply:
client.execute will throw an error if it fails to execute, so return values after it won't be reached, unless I catch errors within each service function. It's simpler to have catches where the service function is called, otherwise we would have a chain of return condition checks for each service function call that catches errors

Reply:
Okay

Reply:
The ones that returns something are queries that select, they will have a populated rows attribute for the return. Functions without a return are queries that insert or update

Reply:
ok